plugin/evm/mempool.go

```diff
@@ -0,0 +1,210 @@
+// (c) 2019-2020, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package evm
+
+import (
+	"sync"
+
+	"github.com/ava-labs/avalanchego/cache"
+	"github.com/ava-labs/avalanchego/ids"
+)
+
+// Mempool is a simple mempool for atomic transactions
+type Mempool struct {
+	lock sync.Mutex
+	// maxSize is the maximum number of transactions allowed to be kept in mempool
+	maxSize int
+	// currentTx is the transaction about to be added to a block.
+	currentTx *Tx
+	// txs is the set of transactions that need to be issued into new blocks
+	txs map[ids.ID]*Tx
+	// issuedTxs is the set of transactions that have been issued into a new block
+	issuedTxs map[ids.ID]*Tx
+	// discardedTxs is an LRU Cache of recently discarded transactions
+	discardedTxs *cache.LRU
+	// Pending is a channel that signals when the mempool is ready to put an atomic transaction
+	// into a new block
+	Pending chan struct{}
```

Comment:
What is the exact invariant relating to this channel?

Reply:
Updated to: Pending is a channel of length one, which the mempool ensures has an item on it as long as there is an unissued transaction remaining in [txs]. This is a little jank because of how it interacts with the awaitSubmittedTxs goroutine. Gonna try to refactor that part to simplify this flow.